source/Nevermore/Util/DataModificationQueryBuilder.cs

```diff
@@ -330,87 +325,177 @@ CommandParameterValues GetDocumentParameters(Func<DocumentMap, string> allocateI
+            {
+                AssertAllowedSplitCommands(totalParameters);
+
+                tableVariableName = CreateGlobalTempTable(sb);
+            }
+            else if (HasRelatedDocuments(relatedDocumentData))
+            {
+                sb.AppendLine($"DECLARE {tableVariableName} as TABLE (Reference nvarchar(400), ReferenceTable nvarchar(400))");
+            }
 
             foreach (var data in relatedDocumentData)
             {
                 if (data.Related.Any())
                 {
                     var relatedVariablePrefix = $"{data.TableName.ToLower()}_";
+                    var remaining = data.Related.Length;
+                    var allValueBlocks = data.Related.Select((r, idx) => $"(@{relatedVariablePrefix}{idx}, '{r.relatedTableName}')").ToArray();
 
-                    sb.AppendLine();
-                    sb.AppendLine("DELETE FROM @references");
-                    sb.AppendLine();
-
-                    var valueBlocks = data.Related.Select((r, idx) => $"(@{relatedVariablePrefix}{idx}, '{r.relatedTableName}')");
-                    sb.Append("INSERT INTO @references VALUES ");
-                    sb.AppendLine(string.Join(", ", valueBlocks));
-                    sb.AppendLine();
+                    while (remaining > 0)
+                    {
+                        var batchSize = Math.Min(1000, Math.Min(SqlServerParameterLimit - parameters.Count, remaining));
+                        var valueBlocks = allValueBlocks.Skip(data.Related.Length - remaining).Take(batchSize);
+
+                        sb.Append($"INSERT INTO {tableVariableName} VALUES ");
+                        sb.AppendLine(string.Join(", ", valueBlocks));
+                        sb.AppendLine();
+                        remaining -= batchSize;
+
+                        for (int i = 0; i < batchSize; i++)
+                        {
+                            var paramIndex = i + data.Related.Length - remaining - batchSize;
+                            parameters.Add(relatedVariablePrefix + paramIndex, data.Related[paramIndex].relatedDocumentId);
+                        }
+
+                        if (parameters.Count >= SqlServerParameterLimit && remaining > 0)
+                        {
+                            commands.Add(new PreparedCommand(sb.ToString(), parameters, RetriableOperation.Update, mapping, updateOptions.CommandTimeout));
+                            sb = new StringBuilder();
+                            parameters = parametersCreator();
+                        }
+                    }
 
                     sb.AppendLine($"DELETE FROM [{data.SchemaName}].[{data.TableName}] WHERE [{data.IdColumnName}] = @{IdVariableName}");
-                    sb.AppendLine($"    AND [{data.RelatedDocumentIdColumnName}] not in (SELECT Reference FROM @references)");
+                    sb.AppendLine($"AND [{data.RelatedDocumentIdColumnName}] not in (SELECT Reference FROM {tableVariableName})");
```

Comment:
@pawelpabich You mentioned table valued parameters before, I think we misunderstood how you would use it. This might actually solve our problem. I'm going to give it a go

https://github.com/OctopusDeploy/Nevermore/wiki/Table-Valued-Parameters#bulk-inserts

Reply:
👍

Reply:
Happy to review the PR once you have something working.

Reply:
@pawelpabich code changes are in, I am taking another look at the existing tests because I think the assertions may be too relaxed

Reply:
Also will do some quick benchmarking